src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -40,17 +41,22 @@ import { INotebookControllerManager } from './types';
  */
 @injectable()
 export class NotebookControllerManager implements INotebookControllerManager, IExtensionSyncActivationService {
-    private controllerMapping = new WeakMap<
-        NotebookDocument,
-        { selected: VSCodeNotebookController | undefined; controllers: VSCodeNotebookController[] }
-    >();
-    private findingInProgress = new WeakMap<NotebookDocument, CancellationTokenSource>();
+    // Keep tabs on which controller is selected relative to each notebook document
+    private controllerMapping = new WeakMap<NotebookDocument, VSCodeNotebookController | undefined>();
+
+    // When opening a document, track our find preferred search so that we can cancel if needed
+    private findPreferredInProgress = new WeakMap<NotebookDocument, CancellationTokenSource>();
+
     private readonly _onNotebookControllerSelected: EventEmitter<{
         notebook: NotebookDocument;
         controller: VSCodeNotebookController;
     }>;
 
+    // Promise to resolve when we have loaded our controllers
```

Comment:
This might end up being too simplistic long term. Right now we're just getting the flat list, we don't handle something like installing the python extension and then regenerating the list. That's marked with a TODO for KERNELPUSH and I'd rather handle that later as long as the basics here are working.